### PR TITLE
fix: Debug Test codelens being skipped for LWCs

### PR DESCRIPTION
**Pull request by randi274**
### What does this PR do?
Addresses an open issue where the Debug Test codelens was not working as expected. I encountered this issue when I realized there was a limit to the number of characters allowed, and the codelens may also not work if the total output is too long (such as in the case of a long directory). The secondary issue is fixed with this PR as well.

### What issues does this PR fix or reference?
#4499, #4690, @W-12549783@

### Functionality Before
Debug Test codelens was not stopping in tests for breakpoints.

### Functionality After
Breakpoints are being hit in individual test runs.

### CODE REVIEW

File: packages/salesforcedx-vscode-lwc/src/testSupport/testRunner/testRunner.ts [hunk 1]
```diff
@@ -95,7 +95,7 @@ export class TestRunner {
       runTestsByPathArgs = [];
     }
     const testNamePatternArgs = testName
-      ? ['--testNamePattern', `"${escapeStrForRegex(testName)}"`]
+      ? ['--testNamePattern', `${escapeStrForRegex(testName)}`]
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-lwc/src/testSupport/testRunner/testRunner.ts:98]: Any findings about what changed here to make this _stop_ working? 


a jest unit test around this particular logic could give us more clarity about how/why it's written this way.

**Review by gbockus-sf — APPROVED** [review]
Notes: 
* There is still an issue on windows when test name include a double quote that isn't located on the end of the name
  * GOOD '\"this is a test\"'
  * BAD 'double \" quotes \" for the loss'
  * I'll open a new bug to call out this behaviour but it's a corner case we may not care about 
* Tested from vsix install 
* When debugging from Tests View ‘true’ is printed to the console.
    * If no breakpoint is set then it’s not clear that the tests were run 


Test values for lwc tests 
Mac:  insiders

* ‘A single test’
    * code lense
        * run ✅
        * debug ✅
    * Tests View
        * run ✅
        * debug ✅
* ‘An ugly test !@#$%^&*()_+\?/{}[]’
    * code lense
        * run ✅
        * debug ✅
    * Tests View
        * run ✅
        * debug ✅
* ‘I have quotes in my “test” ’
    * code lense 
        * run ✅
        * debug ✅
    * Tests View
        * run ✅
        * debug ✅
* ‘I have escaped single \‘ quotes \’ in my test ‘
    * code lense
        * run ✅
        * debug ✅
    * Tests View
        * run ✅
        * debug ✅
* ‘I have escaped double \“ quotes \” in my test’
    * code lense
        * run ✅
        * debug ✅
    * Tests View
        * run ✅
        * debug ✅
* ‘\"Should\ render\ component\ with\ Tax\ Receipts\ and\ Statements\"‘ #original ticket
    * code lense
        * run ✅
        * debug ✅
    * Tests View
        * run ✅
        * debug ✅
* Run all tests
    * npm run test:unit
        * run ✅
    * Add debug LWC jest test debug config 
        * debug ✅
    * Tests View
        * All Test
            * Run ✅
        * Test Suite
            * run ✅
            * debug ✅


Windows: vscode

* ‘A single test’
    * code lense
        * run ✅
        * debug ✅
    * Tests View
        * run ✅
        * debug
* ‘An ugly test !@#$%^&*()_+\?/{}[]’
    * code lense
        * run ✅
        * debug✅
    * Tests View
        * run ✅
        * debug✅
* ‘I have quotes in my “test” ’
    * code lense
        * run ❌
        * debug ❌
    * Tests View
        * run  ❌
        * debug ❌
* ‘I have escaped single \‘ quotes \’ in my test ‘
    * code lense
        * run ✅
        * debug ✅
    * Tests View
        * run ✅
        * debug✅
* ‘I have escaped double \“ quotes \” in my test’
    * code lense
        * run ❌
            * \"I have escaped double quotes in my test\"‘ ✅
            * only fails when the escaped “ are not on the ends of the string
        * debug ❌
    * Tests View
        * run ❌
        * debug❌
* ‘\"Should\ render\ component\ with\ Tax\ Receipts\ and\ Statements\"‘ #original ticket
    * code lense
        * run ✅
        * debug ✅
    * Tests View
        * run ✅
        * debug ✅
* Run all tests
    * npm run test:unit
        * run ✅
    * Add debug LWC jest test debug config 
        * debug ✅
    * Tests View
        * All Test
            * Run ✅
        * Test Suite
            * run ✅
            * debug ✅

**Review by gbockus-sf — DISMISSED** [review]
Let's add a jest tests for this and it'll be good to go

File: packages/salesforcedx-vscode-lwc/src/testSupport/testRunner/testRunner.ts (hunk 1 above)
Comment (jeffb-sfdc) [inline packages/salesforcedx-vscode-lwc/src/testSupport/testRunner/testRunner.ts:98]: The variable `testName` is being reused here.  There's one instance declared on75, and another one here (line 97) line.  Though this is in the original code, I think it should be cleaned up and a different name be used.

Comment (jeffb-sfdc) [inline packages/salesforcedx-vscode-lwc/src/testSupport/testRunner/testRunner.ts:98]: There's a spelling typo on line 68.  Though it's in the original code, it would be nice to get that fixed in this PR.

Comment (jeffb-sfdc) [reply packages/salesforcedx-vscode-lwc/src/testSupport/testRunner/testRunner.ts:98]: There appears to be an issue escaping parenthesis as well.  I created a sample LWC test file in my org, and added the following:

```
    // ****************************************
    it('Example 1', () => {
        expect(1).toBe(1);
    });

    it("Example 2", () => {
        expect(1).toBe(1);
    });

    it(`Example 3`, () => {
        expect(1).toBe(1);
    });


    // ****************************************
    it('Example 4 (aka \'four\')', () => {
        expect(1).toBe(1);
    });

    it("Example 5 (aka 'five')", () => {
        expect(1).toBe(1);
    });

    it(`Example 6 (aka 'six')`, () => {
        expect(1).toBe(1);
    });


    // ****************************************
    it('Example 7 (aka "seven")', () => {
        expect(1).toBe(1);
    });

    it("Example 8 (aka (\"eight\")", () => {
        expect(1).toBe(1);
    });

    it(`Example 9 (aka "nine")`, () => {
        expect(1).toBe(1);
    });
```

When ` 'Example 4 (aka \'four\')' ` is passed to`escapeStrForRegex(testName)`, it returns:

"Example 4 \\(aka 'four'\\)"

expected:
"Example 4 (aka 'four')"

Comment (randi274) [reply packages/salesforcedx-vscode-lwc/src/testSupport/testRunner/testRunner.ts:98]: This was more of a "try and find out" and it worked, but I can definitely add a unit test here too

File: packages/salesforcedx-vscode-lwc/src/testSupport/testRunner/testRunner.ts [hunk 2]
```diff
@@ -94,9 +103,7 @@ export class TestRunner {
     } else {
       runTestsByPathArgs = [];
     }
-    const testNamePatternArgs = testName
-      ? ['--testNamePattern', `${escapeStrForRegex(testName)}`]
-      : [];
+    const testNamePatternArgs = getTestNamePatternArgs(testExecutionInfo);
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-lwc/src/testSupport/testRunner/testRunner.ts:106]: dumb q: why not just pass testExecutionInfo.testName here?

Comment (randi274) [reply packages/salesforcedx-vscode-lwc/src/testSupport/testRunner/testRunner.ts:106]: Not dumb - I was actually kind of unclear why this specific check was happening on the name, and it was a little overly verbose to pass this in as a whole:
`'testName' in testExecutionInfo ? testExecutionInfo.testName : undefined;`

Comment (gbockus-sf) [reply packages/salesforcedx-vscode-lwc/src/testSupport/testRunner/testRunner.ts:106]: gotcha..I'm a sicko so I had to go take a look.  Basically the way types are being used here is pretty weird.  It seems clearer to add at typeguard here ala 

```
// Probably could go in types file. 
function isTestCaseInfo(
  testExecutionInfo: TestExecutionInfo
): testExecutionInfo is TestCaseInfo {
  return testExecutionInfo.kind === TestInfoKind.TEST_CASE;
}

export function getTestNamePatternArgs(testName: string) {
  return ['--testNamePattern', `${escapeStrForRegex(testName)}`];
}

...  
const testNamePatternArgs =
       // Is the type a TestCaseInfo and is testName not emptyString
      (isTestCaseInfo(testExecutionInfo) && testExecutionInfo.testName)
        ? getTestNamePatternArgs(testExecutionInfo.testName)
        : [];
```

Comment (gbockus-sf) [reply packages/salesforcedx-vscode-lwc/src/testSupport/testRunner/testRunner.ts:106]: That check was happening b/c the type `export type TestExecutionInfo = TestCaseInfo | TestFileInfo | TestDirectoryInfo;` is kind of silly.  testName only exists on the `TestCaseInfo` type FYI

File: packages/salesforcedx-vscode-lwc/package.json [hunk 3]
```diff
@@ -78,8 +78,9 @@
     "lint:fix": "npm run lint -- --fix",
     "watch": "tsc -watch -p .",
     "clean": "shx rm -rf node_modules && shx rm -rf out && shx rm -rf coverage && shx rm -rf .nyc_output",
-    "test": "npm run test:unit && npm run test:vscode-integration",
-    "test:unit": "node ./node_modules/nyc/bin/nyc.js ./node_modules/mocha/bin/_mocha --recursive out/test/unit --reporter mocha-multi-reporters --reporter-options configFile=../../config/mochaUnitTestsConfig.json",
+    "test": "npm run test:unit && test:unit:mocha && npm run test:vscode-integration",
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-lwc/package.json:81]: no mocha necessary.

File: packages/salesforcedx-vscode-lwc/package.json [hunk 4]
```diff
@@ -78,8 +78,9 @@
...
+    "test:unit": "jest --coverage",
+    "test:unit:mocha": "node ./node_modules/nyc/bin/nyc.js ./node_modules/mocha/bin/_mocha --recursive out/test/unit --reporter mocha-multi-reporters --reporter-options configFile=../../config/mochaUnitTestsConfig.json",
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-lwc/package.json:83]: this one isn't necessary. The unit tests will get run under the jest execution.

File: packages/salesforcedx-vscode-lwc/test/jest/testSupport/testRunner/testRunner.test.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      const testExecutionInfo: TestCaseInfo = {
+        kind: TestInfoKind.TEST_CASE,
+        testType: TestType.LWC,
+        testUri: mockURI,
+        testName
+      };
+      const testPatternArgs = getTestNamePatternArgs(testExecutionInfo);
+      expect(testPatternArgs.length).toEqual(2);
+      expect(testPatternArgs).toContain('--testNamePattern');
+      expect(testPatternArgs).toContain(testName);
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-lwc/test/jest/testSupport/testRunner/testRunner.test.ts:31]: These tests seem like a perfect opportunity to use snapshots for the expects.  It will capture the expected output you can verify in the snaps, and if/when we update the regex to handle the weird cases it will be simpler to update the tests.

File: packages/salesforcedx-vscode-lwc/src/testSupport/types/index.ts [hunk 6]
```diff
@@ -42,6 +42,15 @@ export const enum TestInfoKind {
   TEST_DIRECTORY = 'testDirectory'
 }
 
+/**
+ * Confirms if the TestExecutionInfo is specific to a testCase
+ */
+export function isTestCase(
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-lwc/src/testSupport/types/index.ts:48]: nit: being that you're type checking the `TestCaseInfo` type `isTestCaseInfo` seems like a better name.

File: packages/salesforcedx-vscode-lwc/test/jest/testSupport/testRunner/testRunner.test.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+      const testNameEscaped = 'Test \\?\\$\\^\\*\\(\\)\\.\\[\\]\\{\\}\\|\\+ Symbols';
+      const testPatternArgs = getTestNamePatternArgs(testName);
+      expect(testPatternArgs).toContain(testNameEscaped);
+      expect(testPatternArgs).toMatchSnapshot();
+    });
+    it('Should not escape certain symbols if testNamePattern is included', async () => {
+      const testName = 'Test !@#"%&;:,<>\=~` Symbols';
+      const testPatternArgs = getTestNamePatternArgs(testName);
+      expect(testPatternArgs).toContain(testName);
+      expect(testPatternArgs).toMatchSnapshot();
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-lwc/test/jest/testSupport/testRunner/testRunner.test.ts:31]: nit: So the toMatchSnapshot is checking the same thing as the toContain for all tests.

File: packages/salesforcedx-vscode-lwc/test/jest/testSupport/testRunner/testRunner.test.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    });
+
+  });
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-lwc/test/jest/testSupport/testRunner/testRunner.test.ts:34]: potential other good test cases to have for documenting behavior 
`‘I have escaped double \" quotes \" in my test’`
`‘\"I have escaped double  quotes  in my test\"’
`"Example with escaped parans  \(aka 'four'\)` // from @jeffb-sfdc testing.

File: packages/salesforcedx-vscode-lwc/test/jest/testSupport/testRunner/types.test.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { isTestCase, TestCaseInfo, TestFileInfo, TestInfoKind, TestType } from '../../../../src/testSupport/types';
+import { Uri } from 'vscode';
+import * as path from 'path';
+
+describe('test support types Unit Tests.', () => {
+  const mockUriPath = path.join('/');
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-lwc/test/jest/testSupport/testRunner/types.test.ts:13]: I guess path changes based on OS.

**Review by gbockus-sf — APPROVED** [review]
great stuff.  I did have a couple of nits you can decide if you want to address.

File: packages/salesforcedx-vscode-lwc/test/jest/testSupport/testRunner/testRunner.test.ts (hunk 7 above)
Comment (randi274) [reply packages/salesforcedx-vscode-lwc/test/jest/testSupport/testRunner/testRunner.test.ts:31]: Got it! I actually think I'm going to leave these in just since it's more readable for the explicit thing we're testing, and I like having the snapshot for the making sure it doesn't change

Comment (gbockus-sf) [reply packages/salesforcedx-vscode-lwc/test/jest/testSupport/testRunner/testRunner.test.ts:31]: cool.  It's not that big a deal.  Just don't burn it in as the way to do snapshot tests :).  One downside is if we change the regex to account for the exception cases we know about today the tests will all need to be touched. snapshot tests only would be more an an auto without code changes (jest would update the snaps).
